src/gui/glRenderer.js

```diff
@@ -1,476 +1,753 @@
-                console.log(frame);
-            }
+        let res;
+        // filter vertex array extension
+        if (localCommand.name.startsWith('extVao-')) {
+            let fnName = localCommand.name.split('-')[1]; // e.g. createVertexArrayOES
+            fnName = fnName.replace(/OES$/, '');
+            res = this.gl[fnName].apply(this.gl, localCommand.args);
+        } else {
+            // execute command
+            // we filter the _bufferSize postfix. This postFix informs us to return the size of the result in the responseBuffer, rather than the real result of the executed command.
+            // this is used if we want to return a variable length string in a shared buffer, but we don't know it's size.
+            // for example, the getActiveUniform returns the name of the uniform, before we can send back the name of the uniform, we need to know it's length
+            // the second call wil, be without the _bufferSize postfix and will indicate that the buffer is big enough to fit the result
+            res = this.executeGL(localCommand.name.endsWith("_bufferSize") ? localCommand.name.substring(0, localCommand.name.length - 11) : localCommand.name, localCommand.args, command.args);
         }
 
-        executeFrameCommands() {
-            this.buffering = false;
+        // some state changes can only be done after execution
+        this.lastState.postProcessOneCommand(command);
 
-            let setup = [];
-            if (this.lastUseProgram) {
-                setup.push(this.lastUseProgram);
-            }
-            for (let activeTexture in this.lastTextureBinds) {
-                setup.push({
-                    name: 'activeTexture',
-                    args: [parseInt(activeTexture)],
-                });
-                for (let command of Object.values(this.lastTextureBinds[activeTexture])) {
-                    setup.push(command);
+        // if a command has a handle assigned to a command, the resulting internal object should be stored in the uncloneables list, so it can be resolved to the correct object next time it is referenced.
+        if (localCommand.handle !== null) {
+            this.unclonables.set(localCommand.handle, res);
+        }
+        
+        // some commands have a responsebuffer assigned to them, these are blocking commands, where the client awaits a response form the server on the send commandbuffer
+        // we translate the result into bytes and store them in the SharedArrayBuffer, unblocking the thread happens after the buffer has been processed.
+        // in theory we can execute several commands in succession and fill the assigned SharedArrayBuffer responseBuffers and then unlock the client.
+        if (localCommand.name === "getActiveAttrib_bufferSize") {
+            // send the size of the required buffer instead of the struct itself
+            new Int32Array(localCommand.responseBuffer)[0] = res.name.length + 9;   
```

Comment:
It's unclear what the `9` and other following numbers correspond to

Reply:
each uniform and each attribute shader parameter info contains the following information:
https://developer.mozilla.org/en-US/docs/Web/API/WebGLActiveInfo

name, size and type. the last two are 4 bytes/32-bit long, the name is the string length + zero termination.
I will update the documentation to make this more clear